src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexReplacement.cs

```diff
@@ -209,7 +209,58 @@ public string Replace(Regex regex, string input, int count, int startat)
             {
                 return input;
             }
+            Debug.Assert(count > 0);
 
+            // Handle the common case of a left-to-right pattern with no backreferences in the replacement pattern such that the replacement is just a string of text.
+            if (!regex.RightToLeft && !_hasBackreferences)
+            {
+                // With no backreferences, there should either be no rules (in the case of an empty replacement)
+                // or one rule (in the case of a single text string).
+                Debug.Assert(_rules.Length <= 1);
+                Debug.Assert(_rules.Length == 0 || (_rules[0] == 0 && _strings.Length == 1));
+
+                return ReplaceSimpleText(regex, input, _rules.Length != 0 ? _strings[0] : "", count, startat);
+            }
+            else
+            {
+                return ReplaceNonSimpleText(regex, input, count, startat);
+            }
+        }
+
+        private static string ReplaceSimpleText(Regex regex, string input, string replacement, int count, int startat)
+        {
+            // As the replacement text is the same for every match, we don't need to store any information in the segments
+            // about the string itself.  For segments from the input, we store the relevant offset and count.  For segments
+            // that are the replacement text, we store -1 for the offset and count, just serving as a placeholder.
+            var state = (input, replacement, segments: OffsetCountStringBuilder.Create(), inputMemory: input.AsMemory(), prevat: 0, count);
+
+            regex.RunAllMatchesWithCallback(input, startat, ref state, (ref (string input, string replacement, OffsetCountStringBuilder segments, ReadOnlyMemory<char> inputMemory, int prevat, int count) state, Match match) =>
+            {
+                state.segments.Add(state.prevat, match.Index - state.prevat); // the input text from the last match ending until here
+                state.prevat = match.Index + match.Length;
+                if (replacement.Length != 0) // using Replace to remove is common, so special-case the empty string
+                {
+                    state.segments.Add(-1, -1); // placeholder for the replacement
+                }
```

Comment:
Would it make a difference if you stored a single int instead of 2?
E.g., positive n = copy n bytes from input, negative n = skip n bytes from input and add replacement
And then have `ToString` can keep a running total of the offset into input

Or store `(int matchOffset, int matchLength)` without the replacement placeholder, and the `ToString` can do the math to figure out how much to copy?

Reply:
> Would it make a difference if you stored a single int instead of 2?
E.g., positive n = copy n bytes from input, negative n = skip n bytes from input and add replacement
And then have ToString can keep a running total of the offset into input

That'd be nice, though I'm not following how that would work. The offset into the input for a match can't be reverse engineered from that info, if I've understood correctly.

> Or store (int matchOffset, int matchLength) without the replacement placeholder, and the ToString can do the math to figure out how much to copy

That should be feasible. I don't expect it to help much with throughput, but it will mean renting smaller arrays, which is nice.

I'll also switch to just using arrays of ints, as they'll be more reusable by the rest of the app.

Reply:
Done